crypto/hashtable/hashtable.c

```diff
@@ -0,0 +1,682 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ *
+ *
+ *
+ * Notes On hash table design and layout
+ * This hashtable uses a hopscotch algorithm to do indexing.  The data structure
+ * looks as follows:
+ *
+ *   hash          +--------------+
+ *   value+------->+ HT_VALUE     |
+ *      +          +--------------+
+ *  +-------+
+ *  |       |
+ *  +---------------------------------------------------------+
+ *  |       |       |       |       |                         |
+ *  | entry | entry | entry | entry |                         |
+ *  |       |       |       |       |                         |
+ *  +---------------------------------------------------------+
+ *  |                               |                         |
+ *  |                               |                         |
+ *  +---------------------------------------------------------+
+ *  |              +                             +            +
+ *  |        neighborhood[0]               neighborhood[1]    |
+ *  |                                                         |
+ *  |                                                         |
+ *  +---------------------------------------------------------+
+ *                              |
+ *                              +
+ *                         neighborhoods
+ *
+ * On lookup/insert/delete, the items key is hashed to a 64 bit value
+ * and the result is masked to provide an index into the neighborhoods
+ * table.  Once a neighborhood is determined, an in-order search is done
+ * of the elements in the neighborhood indexes entries for a matching hash
+ * value, if found, the corresponding HT_VALUE is used for the respective
+ * operation.  The number of entries in a neighborhood is determined at build
+ * time based on the cacheline size of the target CPU.  The intent is for a
+ * neighborhood to have all entries in the neighborhood fit into a single cache
+ * line to speed up lookups.  If all entries in a neighborhood are in use at the
+ * time of an insert, the table is expanded and rehashed.
+ */
+
+#include <string.h>
+#include <internal/refcount.h>
+#include <internal/rcu.h>
+#include <internal/hashtable.h>
+#include <openssl/rand.h>
+
+#include "internal/numbers.h"
+
+static ossl_unused uint64_t fnv1a_hash(uint8_t *key, size_t len)
```

Comment:
Ought we second guess the callers' hash function?
Sure this is generally good but is it optimal?

Reply:
Agreed, we should not be doing hashing ourselves. We should let the instantiator provide the hash function and provide some utility functions which provide sensible defaults for hashing integers or buffers.

Reply:
Why do you say we're second guessing a callers hash funciton? This function is only used if the caller passes in a NULL pointer for the hash function - i.e. this is the default case where the caller doesn't particularly care how a key gets hashed.